File: 프로그래머스/2/142085. 디펜스 게임/디펜스 게임.py
import heapq


def solution(n, k, enemy):
    answer = 0
    q = []
    
    for i in range(len(enemy)):
        heapq.heappush(q, enemy[i])
        
        if len(q) > k:
            n -= heapq.heappop(q)

        if n < 0:
            answer = i
            break
    
    if answer == 0:
        answer = len(enemy)
        
    return answer

File: 프로그래머스/2/142085. 디펜스 게임/디펜스 게임.cpp
#include <string>
#include <vector>
#include <queue>

using namespace std;

int solution(int n, int k, vector<int> enemy) {
    int answer = 0;
    priority_queue<int> q;
    
    for(int i = 0; i<enemy.size(); i++){
        q.push(-enemy[i]);
        
        if (q.size() > k){
            int x = q.top();
            q.pop();
            n += x;
        }
        
        if (n < 0){
            answer = i;
            break;
        }
    }
    
    if (answer == 0){
        answer = enemy.size();
    }
    
    return answer;
}